community/codegen/src/main/java/org/neo4j/codegen/TryBlock.java

```diff
@@ -19,20 +19,165 @@
     {
-        return emitCatch(exception);
+        CatchState catchState = new CatchState( exception, currentState );
+        this.currentState = catchState;
+        catchStates.add( catchState );
+        return this;
     }
 
     public CodeBlock finallyBlock()
     {
-        return emitFinally();
+        if ( finallyState != null )
+        {
+            throw new IllegalStateException( "Cannot have more than one finally block" );
+        }
+        finallyState = newState( currentState );
+        currentState = finallyState;
+        return this;
+    }
+
+    @Override
+    protected void emit( Consumer<MethodEmitter> emitFunction )
+    {
+        currentState.addAction( emitFunction );
+    }
+
+    @Override
+    public void close()
+    {
+        State nextState = currentState.nextState();
+        if ( nextState == null )
+        {
+            super.emit( ( e ) -> e.tryCatchBlock( tryActions(), catchClauses(), finallyActions(), localVariables, resources ) );
+            super.close();
+        }
+        currentState = nextState;
+
+    }
+
+    @Override
+    protected void endBlock()
+    {
+        //do nothing
+    }
+
+    private List<Consumer<MethodEmitter>> tryActions()
+    {
+        return tryState.actions();
+    }
+
+    private List<CatchClause> catchClauses()
+    {
+        return catchStates.stream().map( c -> new CatchClause( c.exception, c.actions) )
+                .collect( Collectors.toList() );
+    }
+
+    private List<Consumer<MethodEmitter>> finallyActions()
+    {
+        return finallyState == null ? Collections.emptyList() : finallyState.actions();
+    }
+
+    /**
+     * A Try block can be in different states depending on if we're in try, catch or finally.
+     * When nextState returns <tt>null</tt> it means we have no more blocks left and should close
+     * the underlying code block.
+     */
+    private interface State
+    {
+
+        void addAction( Consumer<MethodEmitter> action );
+
+        List<Consumer<MethodEmitter>> actions();
+
+        State nextState();
+    }
+
+    /**
+     * Catch states are special states that stores what kind of exception we are catching.
+     */
+    private static class CatchState implements State
+    {
+
+        private final List<Consumer<MethodEmitter>> actions = new LinkedList<>();
+        private final Parameter exception;
+        private final State parent;
+
+        CatchState( Parameter exception, State parent )
+        {
+            this.exception = exception;
+            this.parent = parent;
+        }
+
+        @Override
+        public void addAction( Consumer<MethodEmitter> action )
+        {
+            actions.add( action );
+        }
+
+        @Override
+        public List<Consumer<MethodEmitter>> actions()
+        {
+            return actions;
+        }
+
+        @Override
+        public State nextState()
+        {
+            return parent;
+        }
+    }
+    private State newState( State parent )
```

Comment:
This looks like it could be static. If this method is static the anonymous inner class will not keep a reference to instantiating instance of the surrounding class.